### PR TITLE
feat(provider/kubernetes): v1 suport tolerations

### PR DESCRIPTION
**Pull request by ethanfrogers**
adds support for Pod tolerations for V1 server groups and run job.

### CODE REVIEW

**Comment by ethanfrogers** [comment]
### Add Toleration
image

### Details
image

**Comment by ethanfrogers** [comment]
@sbwsg @danielpeach i want to tread `operators` and `effects` as constants but `eslint` is complaining that those aren't used in my terminal. any suggestions on how to work around that?

**Comment by ethanfrogers** [comment]
@benjaminws FYI

**Comment by sbwsg** [comment]
Given that the fields are only used inside the template string I think the only way to workaround would be to annotate the `const` declarations with lint comments targeting the no-unused-* rules https://eslint.org/docs/rules/no-unused-vars or https://palantir.github.io/tslint/rules/no-unused-variable/ (will vary depending on which rule's warning you're seeing in the terminal).

**Comment by jrsquared** [comment]
If you convert it all to React, then you won't have to hack around it being used only in the template! 😉

File: app/scripts/modules/kubernetes/src/common/tolerations/tolerations.component.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import { IComponentOptions, IController, module } from 'angular';
+
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/common/tolerations/tolerations.component.ts]: Would add 

```
type IOperator = 'Exists' | 'Equal';
type IEffect = 'NoSchedule' | 'PreferNoSchedule' | 'NoExecute';

interface IToleration {
   key: string;
   operator: IOperator;
   effect: IEffect;
   value: string;
}
```

then in the class:

```
public tolerations: ITolerations[];
public operatiors: IOperator[];
public effects: IEffect[];
```

File: app/scripts/modules/kubernetes/src/common/tolerations/tolerations.component.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    this.tolerations.push({
+      key: '',
+      operator: 'Equal',
+      effect: 'NoSchedule',
+      value: '',
+    });
+  }
+
+  public removeToleration(index: any) {
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/common/tolerations/tolerations.component.ts]: s/`index: any`/`index: number`

File: app/scripts/modules/kubernetes/src/common/tolerations/tolerations.component.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    this.tolerations.splice(index, 1);
+  }
+}
+
+class KubernetesTolerationsComponent implements IComponentOptions {
+  public bindings: any = { tolerations: '=' };
+  public controller: any = KubernetesTolerationsController;
+  public controllerAs: any = 'ctrl';
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/common/tolerations/tolerations.component.ts]: Can omit `any`, `controllerAs` will be implicitly typed as a `string` from the assignment.

File: app/scripts/modules/kubernetes/src/help/kubernetes.help.ts [hunk 4]
```diff
@@ -32,6 +32,11 @@ const helpContents: { [key: string]: string } = {
     '(Optional) Either a number or a percentage (1 vs. 10%) representing the maximum number of new pods to deploy during a rollout.',
   'kubernetes.serverGroup.deployment.maxUnavailable':
     '(Optional) Either a number or a percentage (1 vs. 10%) representing the maximum number of pods to destroy during a rollout.',
+  'kubernetes.serverGroup.tolerations.key': `Key is the taint key that the toleration applies to. Empty means match all taint keys. If the key is empty, operator must be Exists; this combination means to match all values and all keys.`,
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/help/kubernetes.help.ts]: Don't need the backtick syntax, but it's also not hurting anything.

**Comment by ethanfrogers** [comment]
@jrsquared i mean...you aren't wrong. 🤣

**Comment by ethanfrogers** [comment]
@sbwsg @jrsquared @danielpeach totally fair. i'll close and rewrite in react!

**Pull request closed without merging**